File: libgomp/testsuite/libgomp.c-c++-common/declare-target-indirect-1.c
/* { dg-do run } */

#pragma omp begin declare target indirect
int foo(void) { return 5; }
int bar(void) { return 8; }
int baz(void) { return 11; }
#pragma omp end declare target

int main (void)
{
  int x;
  int (*foo_ptr) (void) = &foo;
  int (*bar_ptr) (void) = &bar;
  int (*baz_ptr) (void) = &baz;
  int expected = foo () + bar () + baz ();

#pragma omp target map (to: foo_ptr, bar_ptr, baz_ptr) map (from: x)
  x = (*foo_ptr) () + (*bar_ptr) () + (*baz_ptr) ();

  return x - expected;
}

File: libgomp/testsuite/libgomp.fortran/declare-target-indirect-1.f90
! { dg-do run }

module m
contains
  integer function foo ()
    !$omp declare target to (foo) indirect
    foo = 5
  end function

  integer function bar ()
    !$omp declare target to (bar) indirect
    bar = 8
  end function

  integer function baz ()
    !$omp declare target to (baz) indirect
    baz = 11
  end function
end module

program main
  use m
  implicit none

  integer :: x, expected
  procedure (foo), pointer :: foo_ptr, bar_ptr, baz_ptr

  foo_ptr => foo
  bar_ptr => bar
  baz_ptr => baz

  expected = foo () + bar () + baz ()

  !$omp target map (to: foo_ptr, bar_ptr, baz_ptr) map (from: x)
    x = foo_ptr () + bar_ptr () + baz_ptr ()
  !$omp end target

  stop x - expected
end program
